### PR TITLE
improve the CryptoSetup interface

### PR DESCRIPTION
**Pull request by marten-seemann**
Removes race conditions in the `cryptoSetupClient` and simplifies how packets are sealed.

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -222,6 +186,52 @@ func (c *client) handlePacket(remoteAddr net.Addr, packet []byte) error {
...
+		// check if the server sent the offered version in supported versions
+		if v == c.version {
+			return qerr.Error(qerr.InvalidVersionNegotiationPacket, "Server already supports client's version and should have accepted the connection.")
+		}
+		if v != protocol.VersionUnsupported {
+			hasCommonVersion = true
+		}
+	}
+	if !hasCommonVersion {
+		utils.Infof("No common version found.")
```
Comment (lucas-clemente) [inline client.go]: This log doesn't really offer much infos. Should we remove it?

File: handshake/crypto_setup_client.go [hunk 2]
```diff
@@ -92,7 +92,10 @@ func (h *cryptoSetupClient) HandleCryptoStream() error {
 		}
 
 		// send CHLOs until the forward secure encryption is established
-		if h.forwardSecureAEAD == nil {
+		h.mutex.Lock()
```
Comment (lucas-clemente) [inline handshake/crypto_setup_client.go]: This could be a RLock & RUnlock.

File: handshake/interface.go [hunk 3]
```diff
@@ -0,0 +1,19 @@
+package handshake
+
+import "github.com/lucas-clemente/quic-go/protocol"
+
+// Sealer seals a packet
+type Sealer func(dst, src []byte, packetNumber protocol.PacketNumber, associatedData []byte) []byte
```
Comment (lucas-clemente) [inline handshake/interface.go]: I think I would prefer to define a Sealer interface with a single Seal function. Then we can just return the implementations of the various Sealers directly. If we need a lambda somewhere (although I don't see any place right now), we can to something similar to how https://golang.org/pkg/net/http/#HandlerFunc works. What to you think?

Comment (lucas-clemente) [inline handshake/interface.go]: Actually, we could even split up crypto.AEAD into a Sealer and a Opener and reuse the Sealer interface from there (or embed it in a handshake.Sealer).

File: packet_packer.go [hunk 4]
```diff
@@ -78,10 +78,20 @@ func (p *packetPacker) PackPacket(stopWaitingFrame *frames.StopWaitingFrame, con
 func (p *packetPacker) packPacket(stopWaitingFrame *frames.StopWaitingFrame, leastUnacked protocol.PacketNumber, packetToRetransmit *ackhandler.Packet) (*packedPacket, error) {
 	// packetToRetransmit is only set for handshake retransmissions
 	isHandshakeRetransmission := (packetToRetransmit != nil)
```
Comment (lucas-clemente) [inline packet_packer.go]: I think I made this comment previously: This would be much clearer with s/packetToRetransmit/handshakePacketToRetransmit/g ;)

File: client.go [hunk 5]
```diff
@@ -222,6 +186,52 @@ func (c *client) handlePacket(remoteAddr net.Addr, packet []byte) error {
...
 
+func (c *client) handlePacketWithVersionFlag(hdr *PublicHeader) error {
+	var hasCommonVersion bool // check if we're supporting any of the offered versions
+	for _, v := range hdr.SupportedVersions {
+		// check if the server sent the offered version in supported versions
+		if v == c.version {
+			return qerr.Error(qerr.InvalidVersionNegotiationPacket, "Server already supports client's version and should have accepted the connection.")
+		}
+		if v != protocol.VersionUnsupported {
+			hasCommonVersion = true
```
Comment (marten-seemann) [inline client.go]: True. But this doesn't really belong in this PR.

File: client.go [hunk 6]
```diff
@@ -151,67 +151,31 @@ func (c *client) handlePacket(remoteAddr net.Addr, packet []byte) error {
 	}
 	hdr.Raw = packet[:len(packet)-r.Len()]
 
+	c.mutex.Lock()
```
Comment (marten-seemann) [inline client.go]: The defer would include the `session.handlePacket`.

Comment (lucas-clemente) [inline client.go]: Which should be fine, since that only pushed the packet into the chan.

Comment (marten-seemann) [inline client.go]: You're right. I'll change it.

File: client.go (hunk 1 above)
Comment (marten-seemann) [inline client.go]: I'll make a separate commit for the version stuff after we merge this.

Comment (lucas-clemente) [inline client.go]: sgtm
